### PR TITLE
Windows preparsed data dump 

### PR DESCRIPTION
**Pull request by JoergAtGithub**
Now based on the get-descriptor branch

### CODE REVIEW

File: windows/pp_data_dump/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    list(APPEND HIDAPI_pp_data_dump_TARGETS pp_data_dump)
+else()    
+      message(FATAL_ERROR "Can only work on Windows ")
+endif()
+
+install(TARGETS ${HIDAPI_pp_data_dump_TARGETS}
+    RUNTIME DESTINATION "${CMAKE_INSTALL_BINDIR}"
+)
+
+hidapi_configure_pc("../../pc/hidapi.pc.in")
```
Comment (Youw) [inline windows/pp_data_dump/CMakeLists.txt:23]: I believe it is not needed

File: windows/pp_data_dump/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,23 @@
+project(pp_data_dump C)
+
+set(HIDAPI_pp_data_dump_TARGETS)
+if(WIN32)
+    add_executable(pp_data_dump pp_data_dump.c)    
+    set_target_properties(hid_report_reconstructor_test
+        PROPERTIES
+            C_STANDARD 11
+            C_STANDARD_REQUIRED TRUE
+    )
```
Comment (Youw) [inline windows/pp_data_dump/CMakeLists.txt:10]: did you mean:
```suggestion
    set_target_properties(pp_data_dump
        PROPERTIES
            C_STANDARD 11
            C_STANDARD_REQUIRED TRUE
    )
```
?

File: windows/pp_data_dump/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,23 @@
+project(pp_data_dump C)
+
+set(HIDAPI_pp_data_dump_TARGETS)
+if(WIN32)
```
Comment (Youw) [inline windows/pp_data_dump/CMakeLists.txt:4]: lets keep CMake here as simple as possible

remove the variable and the condition
none of that is needed - this target only included on Windows and will fail to build on non-windows platforms anyway

File: windows/pp_data_dump/.gitignore [hunk 4]
```diff
@@ -0,0 +1,15 @@
+Debug
```
Comment (Youw) [inline windows/pp_data_dump/.gitignore:1]: is this file really needed?

File: CMakeLists.txt [hunk 5]
```diff
@@ -65,6 +65,12 @@ if(HIDAPI_WITH_TESTS)
     enable_testing()
 endif()
 
+if(WIN32)
+    option(HIDAPI_BUILD_PP_DATA_DUMP "Build small Windows console application pp_data_dump.exe" ${IS_DEBUG_BUILD})
+else()
+    set(HIDAPI_BUILD_PP_DATA_DUMP OFF)
```
Comment (Youw) [inline CMakeLists.txt:71]: no CMake code outside `windows/CMakeLists.txt` would use it, so the entire wariable outside `if(WIN32)` is needed
```suggestion
```

**Comment by DJm00n** [comment]
@JoergAtGithub I have uploaded preparsed data dumps here: https://github.com/DJm00n/ControllersInfo/commit/c6a434e430df7039ca1531f97f785d68bbfa2c7c

File: windows/pp_data_dump/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,15 @@
+project(pp_data_dump C)
+
+add_executable(pp_data_dump pp_data_dump.c)    
+set_target_properties(pp_data_dump
+    PROPERTIES
+        C_STANDARD 11
+        C_STANDARD_REQUIRED TRUE
+)
+target_link_libraries(pp_data_dump
+PRIVATE hidapi_include hidapi_winapi
```
Comment (Youw) [inline windows/pp_data_dump/CMakeLists.txt:10]:
```suggestion
    PRIVATE hidapi_winapi
```

File: windows/pp_data_dump/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+set_target_properties(pp_data_dump
+    PROPERTIES
+        C_STANDARD 11
+        C_STANDARD_REQUIRED TRUE
+)
+target_link_libraries(pp_data_dump
+    PRIVATE hidapi_winapi
+)
+
+install(TARGETS ${HIDAPI_pp_data_dump_TARGETS}
```
Comment (Youw) [inline windows/pp_data_dump/CMakeLists.txt:13]:
```suggestion
install(TARGETS pp_data_dump
```

File: windows/CMakeLists.txt [hunk 8]
```diff
@@ -49,3 +49,7 @@ hidapi_configure_pc("${PROJECT_ROOT}/pc/hidapi.pc.in")
 if(HIDAPI_WITH_TESTS)
      add_subdirectory(test)
 endif()
+
+if(HIDAPI_BUILD_PP_DATA_DUMP)
+    add_subdirectory(pp_data_dump)
```
Comment (Youw) [inline windows/CMakeLists.txt:54]: Lets add `pp_data_dump/README.md` and briefly describe what is it for and suggested usage.

**Review by Youw — APPROVED** [review]
Other than README - all looks good.

File: windows/pp_data_dump/README.md [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+<vendor_id>_<product_id>_<usage>_<usage_table>.pp_data
+```
+for each top-level collection, of each connected HID device.
+
+
+## File content
+
+The content of such a .pp_data file looks like the struct, which HIDAPI uses internally to represent the Preparsed Data.
+
+*Note: Microsoft doesn't document this structure, the original might differ in some detail.*
```
Comment (Youw) [inline windows/pp_data_dump/README.md:19]:
```suggestion
*NOTE:
Windows parses HID report descriptors into opaque `_HIDP_PREPARSED_DATA` objects.
The internal structure of `_HIDP_PREPARSED_DATA` is reserved for internal system use.\
Microsoft doesn't document this structure. [hid_preparsed_data.cc](https://chromium.googlesource.com/chromium/src/+/73fdaaf605bb60caf34d5f30bb84a417688aa528/services/device/hid/hid_preparsed_data.cc) is taken as a reference for its parsing.*
```

File: windows/CMakeLists.txt [hunk 10]
```diff
@@ -49,3 +49,7 @@ hidapi_configure_pc("${PROJECT_ROOT}/pc/hidapi.pc.in")
 if(HIDAPI_WITH_TESTS)
      add_subdirectory(test)
 endif()
+
+if(HIDAPI_BUILD_PP_DATA_DUMP)
```
Comment (Youw) [inline windows/CMakeLists.txt:53]:
```suggestion
if(DEFINED HIDAPI_BUILD_PP_DATA_DUMP AND HIDAPI_BUILD_PP_DATA_DUMP)
```

**Review by Youw — APPROVED** [review]

**Review by Youw — APPROVED** [review]
